security/pkg/server/ca/authenticate/xfcc_authenticator.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package authenticate
+
+import (
+	"fmt"
+	"net/http"
+
+	"google.golang.org/grpc/metadata"
+
+	"github.com/alecholmes/xfccparser"
+
+	"istio.io/istio/pkg/security"
+)
+
+const (
+	XfccAuthenticatorType = "XfccAuthenticator"
+)
+
+// XfccAuthenticator extracts identities from Xfcc header.
+type XfccAuthenticator struct{}
+
+var _ security.Authenticator = &XfccAuthenticator{}
+
+func (xff XfccAuthenticator) AuthenticatorType() string {
+	return XfccAuthenticatorType
+}
+
+// Authenticate extracts identities from Xfcc Header.
+func (xff XfccAuthenticator) Authenticate(ctx *security.AuthContext) (*security.Caller, error) {
+	// First check if client is trusted client so that we can "trust" the Xfcc Header.
+	delegated := &CidrAuthenticator{}
+	if !delegated.CanTrustCaller(ctx) {
+		return nil, fmt.Errorf("caller is not in the trusted network. XfccAuthenticator can not be used")
+	}
+	meta, ok := metadata.FromIncomingContext(ctx.RequestContext)
+
+	if !ok || len(meta.Get(xfccparser.ForwardedClientCertHeader)) == 0 {
+		return nil, nil
+	}
+	xfccHeader := meta.Get(xfccparser.ForwardedClientCertHeader)[0]
+	return buildSecurityCaller(xfccHeader)
+}
+
+// AuthenticateRequest validates Xfcc Header.
+func (xff XfccAuthenticator) AuthenticateRequest(req *http.Request) (*security.Caller, error) {
+	xfccHeader := req.Header.Get(xfccparser.ForwardedClientCertHeader)
+	if len(xfccHeader) == 0 {
+		return nil, nil
+	}
+	return buildSecurityCaller(xfccHeader)
+}
+
+func buildSecurityCaller(xfccHeader string) (*security.Caller, error) {
+	clientCerts, err := xfccparser.ParseXFCCHeader(xfccHeader)
+	if err != nil {
+		return nil, fmt.Errorf("error in parsing xfcc header: %v", err)
+	}
+	ids := []string{}
+	for _, cc := range clientCerts {
+		if len(cc.URI) > 0 {
+			ids = append(ids, cc.URI)
```

Comment:
IMO  the final result - the spiffee URI - should be set in the auth context by the xfcc authenticator. The split logic is very hard to evaluate.

Since we return a list of identities - we should probably include the other identities  DNS, email, CN from the client certificate. We don't use them here - but the interface is generic.